File: crates/detrix-lsp/tests/fixtures/transitive_impure.py
"""
Test fixture: Transitive impure function call chain

Call chain: entry_point -> middle_layer -> read_file
The impure open() call is 2 levels deep.
"""


def read_file(path: str) -> str:
    """Impure function - reads from file system."""
    with open(path, "r") as f:
        return f.read()


def middle_layer(path: str) -> str:
    """Middle function - calls impure read_file."""
    content = read_file(path)
    return content.upper()


def entry_point(path: str) -> str:
    """Entry point - transitively calls impure read_file."""
    return middle_layer(path)

File: crates/detrix-lsp/tests/fixtures/go/transitive_impure.go
// Test fixture: Transitive impure function call chain
//
// Call chain: EntryPoint -> MiddleLayer -> ReadFile
// The impure os.ReadFile call is 2 levels deep.
package main

import (
	"os"
	"strings"
)

// ReadFile is an impure function - reads from file system.
func ReadFile(path string) (string, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	return string(content), nil
}

// MiddleLayer is a middle function - calls impure ReadFile.
func MiddleLayer(path string) (string, error) {
	content, err := ReadFile(path)
	if err != nil {
		return "", err
	}
	return strings.ToUpper(content), nil
}

// EntryPoint is the entry point - transitively calls impure ReadFile.
func EntryPoint(path string) (string, error) {
	return MiddleLayer(path)
}
